Add observation to inbound endpoints

artembilan opened the pull request:
<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
Add observation to inbound endpoints

artembilan committed:
* Improve `Observation` logic in the `MessageProducerSupport` and `MessagingGatewaySupport`
* Clean up Javadocs in observation support classes
* Add test for `MessagingGatewaySupport` instrumentation into `IntegrationObservabilityZipkinTests`
* Add docs for `@EnableIntegrationManagement.observationPatterns()`
* The generated docs for metrics and spans looks OK: have both `Gateway` and `Handler` section now

artembilan committed:
* Fix Checkstyle violations

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/EnableIntegrationManagement.java:
```diff
@@ -62,4 +62,12 @@
+	 * @since 6.0
+	 * @see org.springframework.integration.support.utils.PatternMatchUtils#smartMatch(String, String...)
+	 */
+	String[] observationPatterns() default "*";
```
I think I would make the default `""` - observation is quite expensive so I would suggest that users should opt-in.

artembilan replied:
Well, you opted-in, when you have added this `@EnableIntegrationManagement` into your configuration.
If not -there is no any instrumentation at all.
So, in my feeling if you have added such a marker, then you probably would expect everything to be instrumented.
Otherwise it sounds double: add annotation and then add an `*` pattern to enable all 🤷

garyrussell replied:
I doubt whether anyone would ever want to instrument every component in an integration flow with observation. I would expect that, in most cases, they would only enable it on inbound and outbound endpoints.

Also, people might want simple timers (metrics) without the extra overhead of observation. Using observation for just timers is less performant.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/IntegrationManagementConfigurer.java:
```diff
@@ -80,6 +83,8 @@
 
 	private ObjectProvider<ObservationRegistry> observationRegistryProvider;
 
+	private String[] observationPatterns = { "*" };
```
ditto

artembilan committed:
* Fix `IntegrationMBeanExporter` to filter out a `MessageProducer` from sources.
Marking a `MessageProducerSupport` with an `IntegrationInboundManagement`
causes it to be considered as a source for JMX.
Technically it might have a reason since it is indeed a source, but that's fully different story

artembilan committed:
* No observation instrumentation by default

artembilan committed:
* Fix "no observation by default" configuration logic

garyrussell commented on src/reference/asciidoc/metrics.adoc:
```diff
@@ -150,7 +150,13 @@ registry.config().meterFilter(MeterFilter.deny(id ->
+See its javadocs for a pattern matching algorithm.
+
+IMPORTANT: By default, none of the `IntegrationManagement` components are instrumented with an `ObservationRegistry` bean.
+Can be configured as `*` to match all components.
```
```suggestion
Can be configured as `*` to match all components but, typically, patterns to match only inbound and outbound endpoints would be used, to instrument just those components.
```

artembilan replied:
Why would we say that? 
You know I'm still going to add an observation to `MessageChannel` as well. I guess that one would be as a `PRODUCER` kind.
What objections do you have to not observe channels?
FYI in Spring Cloud Sleuth there was spans processing only on channels though...
Thanks

garyrussell replied:
It will kill performance if all the timers are switched from simple timers to observations.

artembilan replied:
Right. I know what you mean about all those extra objects in messages processing, but I'm still going to expose an observation on channels like we do with timers at the moment.
Just to fulfill the picture and realign with whatever we have so far with SC Sleuth.
As we state now in docs: it is up end-user to decide what components to instrument.
Perhaps they will rely on an observation from Apache Kafka or Rabbit MQ on the input side and on the JDBC or HTTP on the outbound.
The Spring Integration will be configured only for channel.

Also: it doesn't matter where we instrument an observation, the garbage collection noise will be the same.

Therefore I don't see a reason in incorporate your suggestion to this sentence in the docs.
